### PR TITLE
feat: Allow users to bookmark variants and genes (#95)

### PR DESCRIPTION
**Pull request by gromdimon**
**TODO:**
- [x] Add models, schemas and api in backend
- [x] Add tests
- [x] Add frontend bookmarks behaviour
- [x] Add frontend bookmark page in profile section
- [x] Test

### CODE REVIEW

File: backend/alembic/versions/435a8e31682e_init_bookmarks.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+"""empty message
```
Comment (holtgrewe) [inline backend/alembic/versions/435a8e31682e_init_bookmarks.py:1]: Please put something sensible here, cf

https://github.com/bihealth/reev/blob/main/backend/alembic/versions/8ccd31a4f116_init_adminmsgs.py#L1

File: backend/alembic/versions/435a8e31682e_init_bookmarks.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table(
+        "bookmarks",
+        sa.Column("id", fastapi_users_db_sqlalchemy.generics.GUID(), nullable=False),
+        sa.Column("user", sa.Uuid(), nullable=False),
```
Comment (holtgrewe) [inline backend/alembic/versions/435a8e31682e_init_bookmarks.py:25]: Should this not be a foreign key with on delete cascade to the user so bookmarks are removed when the user record is removed?

File: backend/app/api/api_v1/endpoints/bookmarks.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        raise HTTPException(status_code=403, detail="Not enough permissions")
+
+
+@router.get("/get-by-id", response_model=schemas.BookmarkRead)
+async def get_bookmark(id: str, db: AsyncSession = Depends(deps.get_db)):
+    """
+    Get a bookmark by id. Available only for superusers.
+
+    **Parameters**
+    * **id**: id of the bookmark
```
Comment (holtgrewe) [inline backend/app/api/api_v1/endpoints/bookmarks.py:57]: Is this the proper pydoc syntax? Can you include the API in sphinx documentation and check?

File: backend/app/crud/bookmarks.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+                self.model.obj_type == obj_type,
+                self.model.obj_id == obj_id,
+            )
+        )
+        result = await session.execute(query)
+        obj = result.scalars().first()
+        if obj:
+            await session.delete(obj)
+            await session.commit()
+        return obj
```
Comment (holtgrewe) [inline backend/app/crud/bookmarks.py:50]: Is it necessary to have a particular CRUD here or would it not be enough to use the default?

File: backend/app/crud/bookmarks.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                self.model.obj_type == obj_type,
+                self.model.obj_id == obj_id,
+            )
+        )
+        result = await session.execute(query)
+        return result.scalars().first()
+
+    async def remove_by_user_and_obj(
+        self, session: AsyncSession, *, user_id: str, obj_type: str, obj_id: str
+    ) -> Bookmark:
```
Comment (holtgrewe) [inline backend/app/crud/bookmarks.py:37]: Why not remove by UUID, can be combined with `get_by_user_and_obj`?

File: frontend/src/api/bookmarks.ts [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+  /**
+   * Obtains the currently logged in user's bookmarks.
+   *
+   * @returns bookmarks list for the current user
+   */
+  async fetchBookmarks(): Promise<any> {
+    // Obtain the current user's information
+    await this.fetchCurrentUser()
+    const url = `${this.apiBaseUrl}bookmarks/list?user_id=${this.currentUserId}`
```
Comment (holtgrewe) [inline frontend/src/api/bookmarks.ts:35]: The backend knows the current user and should only allow access to the current user's bookmarks - or any other entity.

File: frontend/src/api/bookmarks.ts [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+   * Obtains the currently logged in user's bookmark for the given object.
+   *
+   * @param obj_type object type, e.g., "seqvar"
+   * @param obj_id object ID, e.g., "HGNC:1100"
+   * @returns bookmark for the current user
+   */
+  async fetchBookmark(obj_type: string, obj_id: string): Promise<any> {
+    // Obtain the current user's information
+    await this.fetchCurrentUser()
+    const url = `${this.apiBaseUrl}bookmarks/get?user_id=${this.currentUserId}&obj_type=${obj_type}&obj_id=${obj_id}`
```
Comment (holtgrewe) [inline frontend/src/api/bookmarks.ts:54]: same here regarding current user

File: backend/app/crud/bookmarks.py (hunk 5 above)
Comment (holtgrewe) [reply backend/app/crud/bookmarks.py:37]: Backend apparently does not interpret query parameter `user` anyway and uses current user as I would expect.

File: frontend/src/components/BookmarkButton.vue [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  try {
+    const bookmark = await client.fetchBookmark(props.type, props.id)
+    if (bookmark) {
+      isBookmarked.value = true
+    }
+  } catch (e) {
+    console.error(e)
+    isBookmarked.value = false
+  }
+})
```
Comment (holtgrewe) [inline frontend/src/components/BookmarkButton.vue:32]: All client interaction should go through stores, not components.

File: frontend/src/components/BookmarkButton.vue [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  } else {
+    await client.createBookmark(props.type, props.id)
+  }
+  try {
+    await client.fetchBookmark(props.type, props.id)
+    isBookmarked.value = true
+  } catch (e) {
+    isBookmarked.value = false
+  }
+}
```
Comment (holtgrewe) [inline frontend/src/components/BookmarkButton.vue:47]: same here

File: frontend/src/views/ProfileView.vue [hunk 10]
```diff
@@ -18,19 +21,44 @@ const logout = async () => {
...
+ * @param geneSymbol Gene symbol to search for
+ */
+const performSearch = async (geneSymbol: string) => {
+  const routeLocation: any = await search(geneSymbol, 'grch37')
+  if (routeLocation) {
+    router.push(routeLocation)
+  } else {
+    console.error(`no route found for ${geneSymbol}`)
+  }
+}
```
Comment (holtgrewe) [inline frontend/src/views/ProfileView.vue:40]: Is this really gene symbol? Not rather the object identifier / small / large variant?

**Review by holtgrewe — DISMISSED** [review]
Generally looks sane except for some discrepancies in the frontend.

File: backend/alembic/versions/435a8e31682e_init_bookmarks.py (hunk 2 above)
Comment (gromdimon) [reply backend/alembic/versions/435a8e31682e_init_bookmarks.py:25]: Good point!

File: backend/app/crud/bookmarks.py (hunk 4 above)
Comment (gromdimon) [reply backend/app/crud/bookmarks.py:50]: The default CRUD allows getting objects only by id. If we need another functionality (to choose objects by type, and obj_id), then CRUD should be extended. That's exactly what I did.

**Pull request merged**